input-output-hk/daedalus
# [DDW-697] Automate Theme Management

This PR automates the manual maintenance of the existing Daedalus themes through the use of the `createTheme` object and the `themes:update` script. Using this script, all existing themes will be compared to the `createTheme` object, and if any definitions are missing, then themes with missing definitions can be updated with the needed CSS variables and `createTheme` will also programmatically generate the color values for the new CSS vars.

## How to test (for developers)

Since there are no new variables on `createTheme` that need to be written to the existing 3 themes, when you run `yarn themes:update` in the nix-shell you will should see this in the terminal:

<img width="739" alt="Screen Shot 2019-07-16 at 10 03 01 PM" src="https://user-images.githubusercontent.com/13347886/61348720-8bd13000-a816-11e9-8083-861f1ae0cb9b.png">

For the purpose of demoing the `themes:update` script, I've added a CSS variable to `createTheme` called `--theme-button-border-color-demo`: `${border}`

<img width="676" alt="Screen Shot 2019-07-16 at 10 04 54 PM" src="https://user-images.githubusercontent.com/13347886/61348805-d18df880-a816-11e9-8e1b-c4912d742fc1.png">

Running `yarn themes:update` again will now show in the terminal that this variable is missing in the 3 themes and will also show the color value `createTheme` generated for each given theme. It will ask if you want to write this new variable to the themes.

<img width="655" alt="Screen Shot 2019-07-16 at 10 05 37 PM" src="https://user-images.githubusercontent.com/13347886/61348882-0d28c280-a817-11e9-96e5-8cabeb14aad2.png">

After entering `yes`, the script will write the variable and values to each theme file, and then run `prettier` on the file so it is formatted properly. You can check the output by running `git diff`.

<img width="816" alt="Screen Shot 2019-07-16 at 10 08 00 PM" src="https://user-images.githubusercontent.com/13347886/61348955-482af600-a817-11e9-9c0c-339433440309.png">

The demo variable above has not been included in this PR, but is just an example out how this script can be used. It will also be possible to write an entire theme to a file with this script by taking the full output of `createTheme` (with new color params) and writing it to a new theme file alongside our current 3. This will be used in the next PR for `DDW-757` which adds new themes.



## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)

---

## Review Checklist

### Basics


### Code Quality

### Testing

### After Review
